### DISCUSSION TITLE
Support for SQLModel

### DISCUSSION THREAD

**Discussion by caniko** [fastapi-users/fastapi-users#1202] [Q&A] [answered]
Hello again @frankie567, thank you for such a quick response yesterday. I really appreciate it!

I know you have said that you won't support SQLModel, but have you reconsidered lately? Nonetheless, looking to move away from `tortoise`, and SQLModel seems like a great choice.

Do you recommend against using the [old adapter](https://github.com/fastapi-users/fastapi-users-db-sqlmodel)? Why don't you merge the version bump PR?

**Comment by frankie567** [comment, marked as answer]
I don't think I'll reconsider SQLModel for now.

I've built (and continue to maintain) apps using SQLModel and I would say that my experience is mixed. Having only one way to define both Pydantic and ORM models sounds appealing at first, but I find the class architecture/inheritance pattern quite messy, in particular if we need to have several "read"/"update"/"create" models.

Besides, when you have more advanced cases (like cascade constraints, hybrid properties, etc), you'll quickly need to surface SQLAlchemy tools and settings.

I'm much more comfortable using SQLAlchemy ORM directly and use it to define my DB models on one side, and have my Pydantic schemas for the API on the other side. Since SQLAlchemy 2, the support for typing and asyncio are now very mature and well integrated. As such, the need for SQLModel is less obvious in my opinion.

If you're looking for a new ORM, I would suggest you take a look at SQLAlchemy. Their documentation may be dreadful, but this is a really nice piece of tech that supports a wide range of use cases: https://docs.sqlalchemy.org/en/20/orm/quickstart.html

**Reply by UnoYakshi** [reply]
Can't say I agree with the answer completely. The main reason I've gotten into SQLModel in the first place is to reduce the size of repeated code base. I also find read/update/create models separation really useful (and easy to understand) when you want to skip some fields, for example.

The problems started when I had some SQLAlchemy (ver. 1.4) models that I couldn't translate into SQLModel paradigm so easily. My example was an external dependencies mixin which broke the new SQLModel-based adapter.
Some stuff still require transmitting extra arguments (`sa_args`/`sa_kwargs`) — doesn't look neat yet still the code base was smaller.

Or... there is a chance I don't know how to cook SQLAlchemy ORM + Pydantic correctly, w/o duplicating SA models' fields into Pydantic ones.

**Reply by UnoYakshi** [reply]
Should I drop SQLModel and switch to SQLAlchemy 2.0? Can I somehow extract the models' schemas to Pydantic?

**Reply by caniko** [reply]
> Or... there is a chance I don't know how to cook SQLAlchemy ORM + Pydantic correctly, w/o duplicating SA models' fields into Pydantic ones.

No, SQLAlchemy only supports `dataclass`, not Pydantic BaseModel. That is one of the reasons SQLModel, which uses SQLAlchemy under the hood, was created in the first place.

> The main reason I've gotten into SQLModel in the first place is to reduce the size of repeated code base. I also find read/update/create models separation really useful (and easy to understand) when you want to skip some fields, for example.

You are right @UnoYakshi, SQLModel and tortoise-orm are much more DRY principled than SQLAlchemy; moreover, frankie suggests that it causes more problems than it solves most of the time. I personally have had experiences with both, and think they are useful tools that should be used at the discretion of the engineer. To say one is *better* than the other is a mistake and a lost opportunity.

If you have no complex interactions it makes sense to use SQLModel; if you do, perhaps SQLAlchemy is better. Blunt vs sharp instrument. Moreover, `fastapi-users` is not made for us, it is made for frankie and his own products. If you don't like it, make your own project, but I must say this is a very good project thanks to his excellent maintenance and authorship.

You will save a lot of time by using fastapi-users, that alone should be a good enough reason to use SQLAlchemy IMHO; unless you like to do it yourself (DIY) for learning and growing.

**Comment by frankie567** [comment]
For your convenience, I've pushed a new release with fixed dependency pins for SQLModel and SQLAlchemy: https://github.com/fastapi-users/fastapi-users-db-sqlmodel/releases/tag/v0.3.0

**Reply by dineshbvadhia** [reply]
It would be helpful to show how to use SQLModel (v0.3.0) with the Full Example (https://fastapi-users.github.io/fastapi-users/11.0/configuration/full-example/) - it might also stop or reduce the questions on this subject ? Thanks :)
